simpleclient_servlet/src/main/java/io/prometheus/client/filter/MetricsFilter.java

```diff
@@ -172,17 +174,37 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
 
         String path = request.getRequestURI();
 
-        String components = getComponents(path);
-        String method = request.getMethod();
-        Histogram.Timer timer = histogram
+        final String components = getComponents(path);
+        final String method = request.getMethod();
+        final Histogram.Timer timer = histogram
             .labels(components, method)
             .startTimer();
 
+        boolean isAsync = false;
         try {
             filterChain.doFilter(servletRequest, servletResponse);
+            isAsync = servletRequest.isAsyncStarted();
+            if (isAsync) {
+                servletRequest.getAsyncContext().addListener(new AsyncListener() {
+                    volatile boolean done = false;
+                    private void meter() {
+                        if (!done) {
+                            done = true;
```

Comment:
This isn't thread safe, because there's a _check-then-act_ race condition. I don't think thread safety is needed here, because my understanding is that `meter()` will only be called once. If it's only called once, please remove `done`. If there are cases where it is called multiple times in different threads, fix the race condition.

Reply:
Makes sense… I can't find anything definitive [on the 3.0 specs](https://download.oracle.com/otndocs/jcp/servlet-3.0-fr-oth-JSpec/) that only one of the three methods will ever be called, so I'll rather err on the safe side by fixing the race.